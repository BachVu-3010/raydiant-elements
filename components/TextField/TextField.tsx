import FormControl from '@material-ui/core/FormControl';
import * as React from 'react';
import FormHelperText from '../FormHelperText';
import Input from '../Input';
import InputBackground from '../InputBackground';
import InputLabel from '../InputLabel';

interface TextFieldProps {
  /** The label of the text field */
  label: string;
  /** The label of the text field */
  value?: string;
  /** The type of input */
  type?: 'text' | 'email' | 'password' | 'tel' | 'search' | 'url';
  /** Set to true to display input with error */
  error?: boolean;
  /** Set to true to disable in the input */
  disabled?: boolean;
  /** Set to true to display as a textarea */
  multiline?: boolean;
  /** Optional helper text */
  helperText?: React.ReactNode;
  /** Optional icon to display */
  icon?: React.ReactNode;
  /** Called when the value changes */
  onChange?: (value: string) => any;
}

class TextField extends React.Component<TextFieldProps, {}> {
  static defaultProps = {
    value: '',
    type: 'text',
    error: false,
    disabled: false,
    multiline: false,
    helperText: '',
    onChange: () => {
      return;
    },
  };

  render() {
    const {
      label,
      value,
      type,
      icon,
      error,
      disabled,
      multiline,
      helperText,
      onChange,
    } = this.props;

    return (
      <FormControl fullWidth error={error}>
        <InputBackground>
          <InputLabel error={error} disabled={disabled}>
            {label}
          </InputLabel>
          <Input
            fullWidth
            value={value}
            type={type}
            disabled={disabled}
            multiline={multiline}
            icon={icon}
            onChange={e => onChange(e.target.value)}
          />
        </InputBackground>
        {helperText && <FormHelperText>{helperText}</FormHelperText>}
      </FormControl>
    );
  }
}

export default TextField;
