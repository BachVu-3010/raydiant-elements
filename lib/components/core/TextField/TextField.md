```js
initialState = {
  text: '',
  email: 'email@example.com',
  password: 'password',
  error: 'Error',
};

<>
  <TextField
    label="Text"
    value={state.text}
    onChange={text => setState({ text })}
  />
  <TextField
    label="Email"
    type="email"
    value={state.email}
    helperText="Please enter your email"
    onChange={email => setState({ email })}
  />
  <TextField
    label="Password"
    type="password"
    value={state.password}
    helperText="Please enter your password"
    onChange={password => setState({ password })}
  />
  <TextField
    error
    label="Error"
    value={state.error}
    helperText="Oops! This field has an invalid value"
    onChange={error => setState({ error })}
  />
</>;
```

### Disabled

```js
<TextField
  label="Text"
  disabled
/>
<TextField
  label="Email"
  type="email"
  value="email@example.com"
  disabled
  helperText="Please enter your email"
/>
<TextField
  label="Password"
  type="password"
  value="password"
  disabled
  helperText="Please enter your password"
/>
<TextField
  error
  label="Error"
  value="Invalid"
  disabled
  helperText="Oops! This field has an invalid value"
/>
```

### Masks

```js
initialState = {
  zip: '',
  phone: '',
  code: '',
};

<>
  <TextField
    label="Zip Code"
    value={state.zip}
    onChange={zip => setState({ zip })}
    mask={[/\d/, /\d/, /\d/, /\d/, /\d/]}
  />
  <TextField
    label="Phone number"
    value={state.phone}
    onChange={phone => setState({ phone })}
    mask={[/\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/, /\d/]}
    maskGuide
  />
  <TextField
    label="Confirmation Code"
    value={state.code}
    mask={(value = '') => {
      const maskArray = [];
      const parts = value.split(/\-/);

      for (let i = 0; i < value.length; i += 1) {
        maskArray.push(/[a-zA-Z\-]/);
      }

      return maskArray;
    }}
    pipe={(value = '') => {
      return value.toUpperCase();
    }}
    onChange={code => setState({ code })}
  />
</>;
```

### Multiline

```js
initialState = {
  text: '',
};

<>
  <TextField
    multiline
    label="Text"
    value={state.text}
    onChange={text => setState({ text })}
  />
  <TextField
    multiline
    label="Text"
    value={state.text}
    helperText="Helper text (optional)"
    onChange={text => setState({ text })}
  />
  <TextField
    multiline
    error
    label="Error"
    value={state.text}
    helperText="Helper text (optional)"
    onChange={text => setState({ text })}
  />
</>;
```
