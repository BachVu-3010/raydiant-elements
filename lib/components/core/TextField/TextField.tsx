import FormControl from '@material-ui/core/FormControl';
import * as React from 'react';
import MaskedInput from 'react-text-mask';
import FormHelperText from '../../internal/FormHelperText';
import Input from '../../internal/Input';
import InputBackground from '../../internal/InputBackground';
import InputLabel from '../../internal/InputLabel';

interface TextFieldProps {
  /** The label of the field */
  label: string;
  /** The value of the field */
  value?: string;
  /** The type of input */
  type?: 'text' | 'email' | 'password' | 'tel' | 'search' | 'url';
  /** Set to true to display input with error */
  error?: boolean;
  /** Set to true to disable in the input */
  disabled?: boolean;
  /** Set to true to display as a textarea */
  multiline?: boolean;
  /** Optional helper text */
  helperText?: React.ReactNode;
  /** Optional icon to display */
  icon?: React.ReactNode;
  /** Optional text mask, only works with type='text' and non-multiline */
  mask?: Array<string | RegExp> | ((value: string) => Array<string | RegExp>);
  /** Set to false to disable the mask guide while typing */
  maskGuide?: boolean;
  /** The mask's placeholder character (defaults to '_') */
  maskPlaceholderChar?: string;
  /** A function to modify the value of a masked input */
  pipe?: (
    conformedValue: string,
    config: any,
  ) => false | string | { value: string; indexesOfPipedChars: number[] };
  /** Called when the value changes */
  onChange?: (value: string) => any;
}

const TextField: React.SFC<TextFieldProps> = ({
  label,
  value = '',
  type = 'text',
  error = false,
  disabled = false,
  multiline = false,
  helperText = '',
  icon = null,
  mask,
  maskGuide = false,
  maskPlaceholderChar = '_',
  pipe,
  onChange = () => {
    return;
  },
}) => {
  const commonProps = {
    type: mask ? 'text' : type,
    multiline: mask ? false : multiline,
    fullWidth: true,
    disabled,
    value,
    icon,
  };
  return (
    <FormControl fullWidth error={error}>
      <InputBackground>
        <InputLabel error={error} disabled={disabled}>
          {label}
        </InputLabel>
        {mask ? (
          <MaskedInput
            placeholder=""
            mask={mask}
            guide={maskGuide}
            placeholderChar={maskPlaceholderChar}
            pipe={pipe}
            onChange={e => onChange(e.target.value)}
            render={(ref, inputProps) => (
              <Input
                {...commonProps}
                inputRef={ref as any}
                inputProps={inputProps}
              />
            )}
          />
        ) : (
          <Input {...commonProps} onChange={e => onChange(e.target.value)} />
        )}
      </InputBackground>
      {helperText && <FormHelperText>{helperText}</FormHelperText>}
    </FormControl>
  );
};

export default TextField;
