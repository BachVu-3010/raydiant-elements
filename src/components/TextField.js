import React from 'react';
import PropTypes from 'prop-types';
import MUITextField from 'material-ui/TextField';
import { createStyleSheet, withStyles } from 'material-ui/styles';
import classnames from 'classnames';

const propTypes = {
  /** Class name(s) */
  className: PropTypes.string,
  /** Whether the button is disabled or not. */
  disabled: PropTypes.bool,
  /** Input should be in an error state.  
   * Pass `true` or `'error'` for an error state.  
   * Pass `'alert'` for an alert state.
  */
  error: PropTypes.oneOf(['alert', 'error', true, false]),
  /** Additional information to help the user fill the field. */
  helperText: PropTypes.string,
  /** Icon associated with the input. */
  // icon: PropTypes.node,
  /** The description for the input field. */
  label: PropTypes.string.isRequired,
  /** Has multiple lines */
  multiline: PropTypes.bool,
  /** Name */
  name: PropTypes.string,
  /** A valid HTML5 input type. */
  type: PropTypes.string,
  /** Ghosted text to display if the input field is empty. */
  placeholder: PropTypes.string,
  /** The value of the input field. */
  value: PropTypes.string,
  /** Called when the user blurs the text field. */
  onBlur: PropTypes.func,
  /** Called when the user modifies the text. */
  onChange: PropTypes.func,
  /** Called when the user focuses the text field. */
  onFocus: PropTypes.func,
  /** @ignore injected by withStyles */
  classes: PropTypes.object.isRequired,
};
const defaultProps = {
  className: '',
  disabled: false,
  error: false,
  helperText: '',
  // icon: null,
  multiline: false,
  name: '',
  placeholder: '',
  type: 'text',
  value: '',
  onBlur: () => {},
  onChange: () => {},
  onFocus: () => {},
};

/**
 * Text field
 */
const TextField = ({
  classes,
  className, disabled, error, helperText, label, multiline, name, placeholder, type, value,
  onBlur, onChange, onFocus }) => {
  let multilineOpts = {};
  if (multiline) {
    multilineOpts = { rows: 4, rowsMax: 4, multiline: true };
  }
  return <MUITextField
    fullWidth
    {...multilineOpts}
    className={classnames(className, classes.root, { [classes.alert]: error === 'alert' })}
    {...{
      disabled,
      error,
      helperText,
      label,
      name,
      placeholder,
      type,
      value,
      onBlur,
      onChange,
      onFocus,
    }}
  />;
};

TextField.propTypes = propTypes;
TextField.defaultProps = defaultProps;

const styleSheet = createStyleSheet(() => ({
  root: {
    '& label': {
      left: '10px',
    },
  },
  alert: {
    '& label': {
      color: '#f8b91c',
    },
    '& > div:after': {
      backgroundColor: '#f8b91c',
      transform: 'scaleX(1)', // alert is always underlined in yellow
    },
  },
}));

export default withStyles(styleSheet)(TextField);
