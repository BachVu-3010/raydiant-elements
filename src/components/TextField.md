##  Single line

```jsx
initialState = {
  city: "San Francisco",
  zip: "",
  password: '',
  verifyPassword: '',
  passwordError: false
};
const onChange = (e) => {
  setState({[e.target.name]: e.target.value});
};
const onZipChange = (e) => {
  const value = e.target.value;
  if (value.length > 5 || !/^\d*$/.test(value)) {
    return;
  }
  onChange(e);
}
const onPassChange = (e) =>  {
  onChange(e);
  const stateNow = {...({password, verifyPassword} = state)};
  stateNow[e.target.name] = e.target.value;
  setState({passwordError: stateNow.password !== stateNow.verifyPassword});
};
const PASSWORD_ERROR = `The "Password" and "Verify Password" fields must match.`;

<Column>
  <TextField name="city" onChange={onChange} value={state.city} label="City" />
  <TextField name="zip" onChange={onZipChange} value={state.zip} placeholder="90210" label="ZIP Code" helperText="Your 5-digit ZIP code. Enter 00000 for non-U.S. addresses." />
  <TextField name="password" type="password" onChange={onPassChange} value={state.password} placeholder="Password" label="Password" error={state.passwordError} />
  <TextField name="verifyPassword" type="password" onChange={onPassChange} value={state.verifyPassword} placeholder="Password, again" label="Verify Password" error={state.passwordError} helperText={state.passwordError ? PASSWORD_ERROR : ''} />
</Column>
```

## Multiline

```jsx
initialState = {value: ""};
const onChange = (e) => {
  setState({value: e.target.value});
};
<div>
<TextField onChange={onChange} value={state.value} label="Multiline" multiline helperText="Type lots of stuff!" />
</div>
```

## States

```jsx
initialState = {error: false, alert: false, disabled: false, value: "Here's a value"};
const onChange = (e) => {
  setState({value: e.target.value});
};

<Column>
  <Row>
    <Checkbox name="error" onChange={e => setState({error: !state.error})} checked={state.error}>Error</Checkbox>
    <Checkbox name="error" onChange={e => setState({alert: !state.alert})} checked={state.alert}>Alert</Checkbox>
    <Checkbox name="disabled" onChange={e => setState({disabled: !state.disabled})} checked={state.disabled}>Disabled</Checkbox>
  </Row>
  <TextField label="Here's a label" value={state.value} onChange={onChange} error={state.alert ? 'alert' : state.error} disabled={state.disabled} />
</Column>
```

## Narrow

```jsx
const aLotOfText = area => `When you have a bunch of ${area} text in a narrow area, it ends up looking a lot like this.`;
<Column style={{width: "300px"}}>
  <TextField value="" label={aLotOfText('label')} />
  <TextField value={aLotOfText('input')} label={aLotOfText('label')} />
  <TextField value={aLotOfText('input')} label="Label" />
  <TextField value="" label="Long helperText" helperText={aLotOfText('helper')} />
  <TextField error value="" label="Long error text" helperText={aLotOfText('error')} />
</Column>
```