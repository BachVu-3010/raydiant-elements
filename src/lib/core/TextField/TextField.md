```js
initialState = {
  text: '',
  email: 'email@example.com',
  password: 'password',
  error: 'Error',
};

<Row inline>
  <TextField
    label="Text"
    value={state.text}
    onChange={text => setState({ text })}
  />
  <TextField
    label="Email"
    type="email"
    value={state.email}
    helperText="Please enter your email"
    onChange={email => setState({ email })}
  />
  <TextField
    label="Password"
    type="password"
    value={state.password}
    helperText="Please enter your password"
    onChange={password => setState({ password })}
  />
  <TextField
    error
    label="Error"
    value={state.error}
    helperText="Oops! This field has an invalid value"
    onChange={error => setState({ error })}
  />
</Row>;
```

### Disabled

```js
<Row inline>
  <TextField label="Text" disabled />
  <TextField
    label="Email"
    type="email"
    value="email@example.com"
    disabled
    helperText="Please enter your email"
  />
  <TextField
    label="Password"
    type="password"
    value="password"
    disabled
    helperText="Please enter your password"
  />
  <TextField
    error
    label="Error"
    value="Invalid"
    disabled
    helperText="Oops! This field has an invalid value"
  />
</Row>
```

### Masks

```js
initialState = {
  zip: '',
  phone: '',
  code: '',
};

<Row inline>
  <TextField
    label="Zip Code"
    value={state.zip}
    onChange={zip => setState({ zip })}
    mask={[/\d/, /\d/, /\d/, /\d/, /\d/]}
  />
  <TextField
    label="Phone number"
    value={state.phone}
    onChange={phone => setState({ phone })}
    mask={[/\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/, /\d/]}
    maskGuide
  />
  <TextField
    label="Confirmation Code"
    value={state.code}
    mask={(value = '') => {
      const maskArray = [];
      const parts = value.split(/\-/);

      for (let i = 0; i < value.length; i += 1) {
        maskArray.push(/[a-zA-Z\-]/);
      }

      return maskArray;
    }}
    pipe={(value = '') => {
      return value.toUpperCase();
    }}
    onChange={code => setState({ code })}
  />
</Row>;
```

### Multiline

```js
initialState = {
  text: '',
};

<Row inline>
  <TextField
    multiline
    label="Text"
    value={state.text}
    onChange={text => setState({ text })}
  />
  <TextField
    multiline
    label="Text"
    value={state.text}
    helperText="Helper text (optional)"
    onChange={text => setState({ text })}
  />
  <TextField
    multiline
    error
    label="Error"
    value={state.text}
    helperText="Helper text (optional)"
    onChange={text => setState({ text })}
  />
</Row>;
```
