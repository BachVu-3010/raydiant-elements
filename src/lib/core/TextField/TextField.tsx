import * as React from 'react';
import InputHelperText from '../InputHelperText';
import InputLabel from '../InputLabel';
import Input from '../Input';
import { IconOptions } from '../Icon';
import useStyles from './TextField.styles';

export interface TextFieldProps {
  /** The label of the field */
  label: string;
  /** The value of the field */
  value?: string;
  /** The type of input */
  type?: 'text' | 'email' | 'password' | 'tel' | 'search' | 'url';
  /** Set to true to display input with error */
  error?: boolean;
  /** Set to true to disable the input */
  disabled?: boolean;
  /** Set the max width */
  maxWidth?: number;
  /** Set the max character length of the input */
  maxLength?: number;
  /** Set to true to display as a textarea */
  multiline?: boolean;
  /** The height of the textarea */
  multilineHeight?: number | string;
  /** Set to true to auto focus the input */
  autoFocus?: boolean;
  /** Optional helper text */
  helperText?: React.ReactNode;
  /** Optional icon to display */
  icon?: React.ReactNode | IconOptions;
  /** Optional text mask, only works with type='text' and non-multiline */
  mask?: Array<string | RegExp> | ((value: string) => Array<string | RegExp>);
  /** Set to false to disable the mask guide while typing */
  maskGuide?: boolean;
  /** The mask's placeholder character (defaults to '_') */
  maskPlaceholderChar?: string;
  /** Preserves the text positions when deleting characters in a masked input */
  keepCharPositions?: boolean;
  /** A function to modify the value of a masked input */
  pipe?: (
    conformedValue: string,
    config: any,
  ) => false | string | { value: string; indexesOfPipedChars: number[] };
  /** Called when the value changes */
  onChange?: (value: string) => any;
  /** Called when the input loses focus */
  onBlur?: React.FocusEventHandler<any>;
  /** Called when the input becomes focused */
  onFocus?: React.FocusEventHandler<any>;
  /** Called when the input is clicked */
  onClick?: (event: React.MouseEvent<any>) => any;
  /** The test id of the input */
  testId?: string;
}

export const TextField: React.SFC<TextFieldProps> = ({
  label,
  value = '',
  type = 'text',
  error = false,
  disabled = false,
  maxWidth,
  maxLength,
  multiline = false,
  multilineHeight,
  autoFocus = false,
  helperText = '',
  icon = null,
  mask,
  maskGuide = false,
  maskPlaceholderChar = '_',
  keepCharPositions = false,
  pipe,
  onChange,
  onBlur,
  onFocus,
  onClick,
  testId,
}) => {
  const classes = useStyles();

  return (
    <div className={classes.root} style={{ maxWidth }}>
      <InputLabel error={error} disabled={disabled}>
        {label}
      </InputLabel>

      <Input
        type={type}
        icon={icon}
        multiline={multiline}
        multilineHeight={multilineHeight}
        mask={mask}
        maskGuide={maskGuide}
        maskPlaceholderChar={maskPlaceholderChar}
        keepCharPositions={keepCharPositions}
        pipe={pipe}
        value={value}
        maxLength={maxLength}
        onBlur={onBlur}
        onFocus={onFocus}
        autoFocus={autoFocus}
        onChange={onChange}
        onClick={onClick}
        disabled={disabled}
        error={error}
        testId={testId}
      />

      {helperText && (
        <InputHelperText indent error={error} disabled={disabled}>
          {helperText}
        </InputHelperText>
      )}
    </div>
  );
};

export default TextField;
