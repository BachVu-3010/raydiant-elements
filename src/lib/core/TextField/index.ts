import TextField, { TextFieldProps } from './TextField';
export { TextFieldProps };
export default TextField;
