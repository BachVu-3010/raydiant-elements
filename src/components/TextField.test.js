import React from 'react';
import { render, mount } from 'enzyme';
import sinon from 'sinon';
import TextField from './TextField';

test('Value passed through', () => {
  const tf = render(<TextField label="label" value="test" />);
  expect(tf.find('input').prop('value')).toEqual('test');
});

test('onChange', () => {
  const onChange = sinon.spy();
  const component = mount(<TextField label="changes" onChange={onChange} />);
  component.find('input').simulate('change');
  expect(onChange.callCount).toEqual(1);
});
