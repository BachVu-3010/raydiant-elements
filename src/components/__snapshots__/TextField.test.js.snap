// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
<TextField
  className=""
  classes={
    Object {
      ".TextField-alert-2 > div:after": "TextField-alert-2 > div:after",
      ".TextField-alert-2 label": "TextField-alert-2 label",
      ".TextField-root-1 label": "TextField-root-1 label",
      "alert": "TextField-alert-2",
      "root": "TextField-root-1",
    }
  }
  disabled={false}
  error={false}
  helperText=""
  label="a label"
  multiline={false}
  name=""
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  placeholder=""
  type="text"
  value="a value"
/>
`;
